src/Inciter/AMR/edge_store.h

```diff
@@ -0,0 +1,329 @@
+                size_t C = face_ids[2];
+
+                edge_t key = nodes_to_key(A,B);
+                key_list[0] = key; // TODO: Is it OK to use copy assignment here?
+
+                key = nodes_to_key(A,C);
+                key_list[1] = key;
+
+                key = nodes_to_key(B,C);
+                key_list[2] = key;
+
+                return key_list;
+            }
+
+            /**
+             * @brief function to take a list of edge and mark them all
+             * as needing to be refined
+             *
+             * @param ids List of ids to mark for refinement
+             */
+            void mark_edges_for_refinement(std::vector<node_pair_t> ids) {
+                for (size_t i = 0; i < ids.size(); i++)
+                {
+                    node_pair_t pair = ids[i];
+                    edge_t key = nodes_to_key(pair[0], pair[1]);
+
+                    mark_for_refinement(key);
+                    trace_out << get(key).needs_refining << std::endl;
+                }
+            }
+
+
+            /**
+             * @brief function to mark a single edge as needing
+             * refinement (provides a nice abstraction from messing with the
+             * struct directly).
+             *
+             * @param key The edge key to mark as refinement
+             */
+            void mark_for_refinement(edge_t key)
+            {
+                    assert( exists(key) );
+                    get(key).needs_refining = true;
+            }
+
+            /**
+             * @brief Function to unmark and edge as needing refinement
+             *
+             * @param key The key representing the edge to unmark
+             */
+            void unmark_for_refinement(edge_t key)
+            {
+                    assert( exists(key) );
+                    get(key).needs_refining = false;
+            }
+
+            // TODO: Document this (and implement!)
+            void mark_edges_for_derefinement(std::vector<node_pair_t> ids) {
+                for (size_t i = 0; i < ids.size(); i++)
+                {
+                    node_pair_t pair = ids[i];
+                    edge_t key = nodes_to_key(pair[0], pair[1]);
+
+                    mark_edge_for_derefinement(key);
+                }
+            }
+            void mark_edge_for_derefinement(edge_t key) {
+                    get(key).needs_derefining = true;
+            }
+
+
+            /**
+             * @brief Function to generate a list of edge keys from a tet
+             *
+             * @param tet The tet to generate edge pairs for
+             *
+             * @return A list (array) of edge keys which can be separated out to
+             * name the two composing node ids
+             */
+            // TODO: Should this return a pointer/reference?
+            edge_list_t generate_keys(tet_t tet)
+            {
+                // FIXME : Generate these with a (2d) loop and not hard code them?
```

Comment:
I would leave it as is. Could probably be done with a loop, but at this low level, I would definitely want that as a compile-time unrolled loop instead of a runtime loop (which the compiler can choose not to unroll). We could look into that but AFAIK it is not trivial, see e.g., https://www.codeproject.com/Articles/857354/Compile-Time-Loops-with-Cplusplus-Creating-a-Gener.